### ISSUE TITLE
fix(graph): resolve missing graph in Correlation view for reports with correlated reports

### ISSUE THREAD

**Issue by dominictory** [LuatixHQ/opencti#9198]
## Description

We have a report created by the CIRCL MISP feed connector. On the Overview page, we see multiple correlated reports:

![Image](https://github.com/user-attachments/assets/838bd8a4-71a3-469b-b9f2-38e953d43ea3)

On the Knowledge tab, we see a graph. However, when attempting to use `Correlation view`, no graph is produced:

![Image](https://github.com/user-attachments/assets/21c77a6f-e1c9-448d-9024-daf56160655b)

## Environment

6.4.0

## Reproducible Steps

- Ingest a report that is correlated with other reports
- Attempt to use `Correlation view` for that report
- Observe no graph is produced

## Expected Output

Graph is produced in `Correlation view` showing correlations

## Actual Output

No graph produced in `Correlation view`. This occurs for other reports as well.

**Comment by nino-filigran** [comment]
@dominictory we know that this area of the app does not work well, that is why we are currentlyworking on a task to align the graph & the data contained in the container overview. https://github.com/OpenCTI-Platform/opencti/issues/3227. 
In short, the 2 queries are different, resulting in a different view, as you have rightfully pointed. Therefore, the feature should fix your bug cc @romain-filigran FYI

**Comment by ckane** [comment]
I think that I have noticed this on graphs with a large number of correlations. I don't know what, if any, limit we might be running into. I suspect that one of the filters applied in the `buildCorrelationData` function has a logic flaw in it.

**Comment by ckane** [comment]
So I added some instrumentation to my deployment and it seems like when the number of relationships goes above 500, this behavior starts to manifest. I suspect that the hard-coded `first: 500` and similar 500-entry limits is being hit somewhere along the way such that the function linked below only gets the first 500 items from the "total objects" list and it ends up not pivoting into ones with related items:
* https://github.com/OpenCTI-Platform/opencti/blob/master/opencti-platform/opencti-front/src/utils/Graph.js#L438

**Comment by romain-filigran** [comment]
@Gwendoline-FAVRE-FELIX @delemaf : Can you check whether your current work on graph and correlation fix this issue ?

**Comment by labo-flg** [comment]
I'm not confident latest changes would fix the issue.
Also, it's very hard to reproduce without the actual dataset (all correlated containers and their content).

IMO the problem comes from multiple stages of limitations (as the one rightfully pointed by @ckane).
The correlation graph was initially developed by taking a straight approach : 

1. query the container, 
2. the content of the container, 
3. for each element in the content, query the containers that also contains this element
4. for each of these containers, query their content too

... all of these in one query.

As containers might be huge (take some MISP reports for instance), we can clearly see where this is going : pulling thousands and thousands of entities from database into your server memory to finally serve the query. This can (and will) lead to an out-of-memory server crash like in https://github.com/OpenCTI-Platform/opencti/issues/9293

Arbitrary limitations are in place, to mitigate this issue. For instance, in step 3 above we limit to the 20 first containers.
... and these leads to inconsistencies and bugs like this one.

The related code base is now very large and complex, in such extent that we must revise the entire design if we do not want to break something while trying to fix something else.

This is what we want to do with https://github.com/OpenCTI-Platform/opencti/issues/9298, planned (as of now) in 6.6 milestone.

**Comment by dominictory** [comment]
> I'm not confident latest changes would fix the issue. Also, it's very hard to reproduce without the actual dataset (all correlated containers and their content).
> 
> IMO the problem comes from multiple stages of limitations (as the one rightfully pointed by [@ckane](https://github.com/ckane)). The correlation graph was initially developed by taking a straight approach :
> 
> 1. query the container,
> 2. the content of the container,
> 3. for each element in the content, query the containers that also contains this element
> 4. for each of these containers, query their content too
> 
> ... all of these in one query.
> 
> As containers might be huge (take some MISP reports for instance), we can clearly see where this is going : pulling thousands and thousands of entities from database into your server memory to finally serve the query. This can (and will) lead to an out-of-memory server crash like in [#9293](https://github.com/OpenCTI-Platform/opencti/issues/9293)
> 
> Arbitrary limitations are in place, to mitigate this issue. For instance, in step 3 above we limit to the 20 first containers. ... and these leads to inconsistencies and bugs like this one.
> 
> The related code base is now very large and complex, in such extent that we must revise the entire design if we do not want to break something while trying to fix something else.
> 
> This is what we want to do with [#9298](https://github.com/OpenCTI-Platform/opencti/issues/9298), planned (as of now) in 6.6 milestone.

Hi, I'm not a programmer so apologies if this won't make sense haha. For very large reports, might it be possible/more realistic to do some kind of 'expand' and 'collapse' function in correlated view, so we can look at correlation when needed, whilst avoiding loading the entire report in one query leading to memory issues? My thinking loosely comes from CrowdStrike, and how when you look at a process tree, only a basic tree of parent processes is loaded whilst giving you the ability to expand child processes if needed to investigate further. So in this case, if we choose to, expand on the different entities/observables in the report, so it isn't just one large query.

**Comment by ckane** [comment]
Interesting idea @dominictory, that's how the Investigation tool works today.

When I did the initial code for the correlation graph awhile back it was definitely just using a naive approach to the problem. I suppose that the platform could implement an upper limit that could detect when a graph would get "too big" (maybe better dynamically adjusted for infrastructure size) and recommended that the user export the report to an investigation, to expand links manually, if that is the case?

**Comment by labo-flg** [comment]
These are good suggestions ! 
We have started preliminary works on a future refactoring of the graphs, and we are taking these discussions into account.
We'll probably change how the data is loading, to do it bits by bits with proper loading feedback in the UI. You would see your graph loading and expanding as new data comes. The goal is to get rid of all these hard-coded limits... and if the graph is "too big" it would just make your browser suffer as it has to handle more and more nodes to display, but the whole platform will still be up and running properly.

I can already tell you this is significant work, the graph code is overly complex and duplicated throughout the platform.
We'll need to clean up this first, and it won't be done in a day...

**Comment by ckane** [comment]
Yeah totally understand, I agree that the duplication of that code throughout the platform absolutely complicates maintainability

**Comment by SamuelHassine** [comment]
🤖 [AI-generated]

Old title: `fix(graph): no graph produced on Correlation view of a report with correlated reports`
New title: `fix(graph): resolve missing graph in Correlation view for reports with correlated reports`

Reason: description must start with a lowercase imperative verb — "no graph produced" is a noun phrase, rewritten to lead with "resolve".
